lib/recomb_map.c

```diff
@@ -196,6 +208,31 @@ recomb_map_phys_to_discrete_genetic(recomb_map_t *self, double x, uint32_t *locu
     return ret;
 }
 
+/* Find the `index` of the interval within `values` the `query` fits, such that
+ * values[index-1] < query <= values[index]
+ * If necessary, will find the leftmost such index
+ */
+size_t
+recomb_map_binary_interval_search(double query, double *values, size_t n_values)
```

Comment:
I'm not sure what we should do at the top of the range here. Probably it's best if we return an error if the query is > values[n_value - 1], otherwise we could have silent errors. Therefore, we'll need to change the return type to ``int64_t`` and return MSP_ERR_BAD_BINARY_SEARCH_QUERY, I think.

Reply:
Should the function be decorated with `MSP_WARN_UNUSED`? In this case, should `recomb_map_genetic_to_phys` return an error code as well?

Reply:
I think that the current implementation also allows the return of the last index at top of the range [source](https://github.com/tskit-dev/msprime/blob/master/lib/recomb_map.c#L236):

```
s = 0;
k = 0;
while (s < x && k < self->size - 1) {
    s += (p[k + 1] - p[k]) * r[k];
    k++;
}
assert(k > 0);
```
if `x` is larger than the last value of `s`, `k` can point to the last element of `p`. This is the behaviour that the `binary_interval_search` has as well (tested [here](https://github.com/tskit-dev/msprime/blob/c596ba848f20473f46043b4a8efed9caf2899630/lib/tests/tests.c#L3119)). This seems consistent to me, as the function looks for an appropriate interval where the query fits. If we look for a very large value, we find the last (open) interval. Maybe I am missing the point though :/

Reply:
No, you're right, let's keep the interface consistent with the existing code.